src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -178,83 +220,158 @@ public static IEnumerable<T> EnumerateAllComponentsInScenesAndPrefabs<T>()
+                        {
+                            lookupByName[assets[i].Name] = assets[i].Ids.ToList();
+                        }
                     }
                 }
-                return lookupByAssetId;
-            }
-        }
 
-        protected IDictionary<TAsset, TAssetEntry> LookupByAsset
-        {
-            get
-            {
-                if (lookupByAsset == null)
-                {
-                    if (assets == null)
-                    {
-                        lookupByAsset = new Dictionary<TAsset, TAssetEntry>();
-                    }
-                    else
-                    {
-                        lookupByAsset = assets.Where(a => a.Asset != null).ToDictionary(a => a.Asset);
-                    }
-                }
-                return lookupByAsset;
+                return lookupByName;
             }
         }
 
+#pragma warning disable 414
+        private Dictionary<string, List<AssetId>> lookupByName;
+        private Dictionary<AssetId, AssetCacheEntry> lookupByAssetId = new Dictionary<AssetId, AssetCacheEntry>();
+        private Dictionary<TAsset, AssetCacheEntry> lookupByAsset = new Dictionary<TAsset, AssetCacheEntry>();
+#pragma warning restore 414
+
         public TAsset GetAsset(AssetId assetId)
         {
             using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "GetAsset"))
             {
-                TAssetEntry assetEntry;
-                if (LookupByAssetId.TryGetValue(assetId, out assetEntry))
+                if (assetId == null ||
+                assetId == AssetId.Empty ||
+                assetId.Name == string.Empty)
+                {
+                    return default(TAsset);
+                }
+
+                if (lookupByAssetId.TryGetValue(assetId, out var entry))
                 {
-                    return assetEntry.Asset;
+                    return (TAsset)entry.Asset;
                 }
-                else
+
+                if (!TryLoadAssets(assetId.Name))
                 {
                     return default(TAsset);
                 }
+
+                if (!lookupByAssetId.TryGetValue(assetId, out entry))
+                {
+                    Debug.LogError($"Assets were loaded for {assetId.Name} but no associated asset cache entry was found for {assetId}");
+                    return default(TAsset);
+                }
+
+                return (TAsset)entry.Asset;
             }
         }
 
         public AssetId GetAssetId(TAsset asset)
         {
             using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "GetAssetId"))
             {
-                TAssetEntry assetEntry;
-                if (asset != null && LookupByAsset.TryGetValue(asset, out assetEntry))
+                if (asset == null ||
+                    asset.name == null ||
+                    asset.name == string.Empty)

```

Comment:
For "simplicity" (or maybe I dunno what the right word is), consider:

[https://docs.microsoft.com/en-us/dotnet/api/system.string.isnullorempty?view=netframework-4.8](https://docs.microsoft.com/en-us/dotnet/api/system.string.isnullorempty?view=netframework-4.8)

I think same for the stuff up above too